cmd/isp/src/protocol.rs

```diff
@@ -0,0 +1,544 @@
+        let r = port.read(&mut frame_bytes[cnt..])?;
+        cnt += r;
+    }
+
+    let frame = FramingPacket::read_from(frame_bytes.as_slice())
+        .ok_or(anyhow!("Parse_err"))?;
+
+    // A response packet is a specific type of command packet.
+    if frame.header.packet_type != (PacketType::Command as u8) {
+        return Err(anyhow!(
+            "Expected a command, got {:x}",
+            frame.header.packet_type
+        ));
+    }
+
+    cnt = 0;
+    let length: usize =
+        (frame.length_low as usize) | ((frame.length_high as usize) << 8);
+    let mut response = vec![0; length];
+
+    while cnt != length {
+        let r = port.read(&mut response[cnt..])?;
+        cnt += r;
+    }
+
+    check_crc(&frame_bytes, &response, &frame)?;
+
+    let command =
+        RawCommand::read_from(&response[..core::mem::size_of::<RawCommand>()])
+            .ok_or(anyhow!("Parse error"))?;
+
+    if command.tag != (response_type as u8) {
+        return Err(anyhow!(
+            "Expected a response type of {:x}, got {:x}",
+            response_type as u8,
+            command.tag
+        ));
+    }
+
+    let mut params: Vec<u32> = Vec::new();
+    let mut cnt = 0;
+    let mut index = core::mem::size_of::<RawCommand>();
+
+    while cnt < command.parameter_count {
+        params.push(u32::from_le_bytes(response[index..index + 4].try_into()?));
+        cnt += 1;
+        index += 4;
+    }
+
+    send_ack(port)?;
+
+    // First paramter is always the return code;
+    let retval = params[0];
+
+    if retval != 0 {
+        // Some more specific error messages.
+        if retval == 10203 {
+            Err(anyhow!("Did you forget to erase the flash? (err 10203)"))
+        } else if retval == 10101 {
+            Err(anyhow!(
+                "Incorrect signature. Is the SBKEK set correctly? (err 10101)"
+            ))
+        } else {
+            Err(anyhow!("ISP error returned: {}", retval))
+        }
+    } else {
+        Ok(params)
+    }
+}
+
+pub fn send_command(
+    port: &mut dyn serialport::SerialPort,
+    cmd: CommandTag,
+    args: Vec<u32>,
+) -> Result<()> {
+    let command = CommandPacket::new_command(cmd, args)?;
+
+    let command_bytes = command.to_bytes()?;
+
+    port.write_all(&command_bytes)?;
+    port.flush()?;
+
+    read_ack(port)?;
+
+    Ok(())
+}
+
+pub fn send_data(
+    port: &mut dyn serialport::SerialPort,
+    data: &[u8],
+) -> Result<()> {
+    let mut cnt = 0;
+
+    // Target doesn't like it when we send an entire binary in one pass
+    // so break it down into 512 byte chunks which is what the existing
+    // tools seem to use
+    while cnt < data.len() {
+        let end = min(data.len(), cnt + 512);
+
+        let data_packet = DataPacket::new_data(data[cnt..end].to_vec())?;
+
+        let data_bytes = data_packet.to_bytes()?;
+
+        port.write_all(&data_bytes)?;
+        port.flush()?;
+
+        read_ack(port)?;
+        cnt += 512;
+    }
```

Comment:
You can use [`chunks`](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.chunks) to avoid doing this by hand:
```suggestion
    // Target doesn't like it when we send an entire binary in one pass
    // so break it down into 512 byte chunks which is what the existing
    // tools seem to use
    for chunk in data.chunks(512) {
        let data_packet = DataPacket::new_data(chunk.to_vec())?;

        let data_bytes = data_packet.to_bytes()?;

        port.write_all(&data_bytes)?;
        port.flush()?;

        read_ack(port)?;
    }
```